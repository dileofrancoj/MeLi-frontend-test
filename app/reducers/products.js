import {
  FETCH_ERROR,
  FETCH_SUCCESS,
  FETCHING,
} from "./../reducers/actions/common";
export const initialState = {
  loading: false,
  error: false,
  products: [],
  categories: [],
};

export function productsReducer(state, action) {
  switch (action.type) {
    case FETCHING:
      return {
        loading: true,
        error: false,
        product: [],
        categorie: [],
      };
    case FETCH_SUCCESS:
      return {
        loading: false,
        error: false,
        products: action.payload.items,
        categories: action.payload.categories,
      };
    case FETCH_ERROR:
      return {
        loading: false,
        error: true,
        products: [],
        categories: [],
      };
    default:
      return {
        state,
      };
  }
}
