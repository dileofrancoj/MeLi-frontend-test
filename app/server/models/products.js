import fetch from "isomorphic-fetch";
import { BASE_URL } from "../utils/constants";

const list = async (q, limit = 4) => {
  const data = await fetch(
    `${BASE_URL}/sites/MLA/search?q=${q}&limit=${limit}`
  );
  return await data.json();
};

const single = async (id) => {
  const data = await fetch(`${BASE_URL}/items/${id}`);
  return await data.json();
};

const singleDescription = async (id) => {
  const data = await fetch(`${BASE_URL}/items/${id}/description`);
  return await data.json();
};

module.exports = { list, single, singleDescription };
