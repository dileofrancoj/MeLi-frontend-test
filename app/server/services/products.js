const sendProducts = (_products) =>
  _products.map(
    ({
      id,
      title,
      currency_id: currency,
      price,
      thumbnail: picture,
      condition,
      shipping,
      address,
    }) => {
      return {
        id,
        title,
        price: {
          currency,
          amount: price,
          decimals: 2,
        },
        picture,
        condition,
        free_shipping: shipping.free_shipping,
        address: address.state_name,
      };
    }
  );

const sendProduct = ({ product: _product, description, categories }) => {
  return {
    id: _product.id,
    title: _product.title,
    price: {
      currency: _product.currency_id,
      amount: _product.price,
      decimals: 2,
    },
    picture: _product.pictures[0].secure_url,
    condition: _product.condition,
    free_shipping: _product.shipping.free_shipping,
    sold_quantity: _product.sold_quantity,
    description,
    categories,
  };
};

module.exports = { sendProducts, sendProduct };
