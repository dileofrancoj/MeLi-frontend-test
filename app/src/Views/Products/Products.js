import React from "react";
/* Bootstrap */
import { Row, Col } from "react-bootstrap";

/* Currency Helper */
import { formatAsCurrency } from "../../../server/utils/helpers";

/* Routing */
import { Link } from "react-router-dom";

const ProductsView = ({
  id,
  title,
  price,
  free_shipping,
  address,
  picture,
}) => {
  return (
    <Row className="justify-content-center bg-white pb-16 pt-16 pl-32 first-row">
      <Col md={9}>
        <Row>
          <Col md={3} className="thumbnail">
            <Link to={`/items/${id}`}>
              <picture>
                <img
                  className="pointer rounded thumbnail"
                  src={picture}
                  alt={title}
                />
              </picture>
            </Link>
          </Col>
          <Col md={9}>
            <h4 className="f-24">
              $ {formatAsCurrency(price.amount)}{" "}
              {free_shipping ? (
                <img
                  src="https://images-prod-meli.s3-sa-east-1.amazonaws.com/ic_shipping.png"
                  alt="free shipping"
                />
              ) : null}
            </h4>

            <Link
              className="mt-32"
              to={`/items/${id}`}
              style={{
                color: "black",
                textDecoration: "inherit",
              }}
            >
              <p className="pointer f-18 ">{title}</p>
            </Link>
          </Col>
        </Row>
      </Col>
      <Col md={3} className="pt-16 pl-16">
        <p className="f-12"> {address}</p>
      </Col>
    </Row>
  );
};

export default ProductsView;
