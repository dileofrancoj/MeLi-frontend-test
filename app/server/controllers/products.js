const { list, single, singleDescription } = require("../models/products");
import { single as singleCat } from "./../models/categories";
const {
  filterToCategories,
  categoriesSingle,
} = require("../services/categories");
const { sendProducts, sendProduct } = require("../services/products");

const author = {
  name: "Franco",
  lastname: "Di Leo",
};

const searchById = async (id) => {
  try {
    const obj = {};
    const product = await single(id);
    const { category_id } = product;
    const { path_from_root } = await singleCat(category_id);

    const categories = filterToCategories(path_from_root);

    const { plain_text: description } = await singleDescription(id);
    // ConfiguraciÃ³n de Babel no me permite usar spread
    Object.assign(obj, { product, description, categories });
    const item = sendProduct(obj);
    return item;
  } catch (e) {
    console.error(e);
  }
};

const search = async (product) => {
  try {
    const { results, filters } = await list(product);
    //if (results.length === 0) return res.sendStatus(404);
    const items = sendProducts(results);
    const categories = filterToCategories(filters);
    return { author, items, categories };
    //res.json({ author, categories, items });
  } catch (e) {
    console.error(e);
  }
};

module.exports = { search, searchById };
