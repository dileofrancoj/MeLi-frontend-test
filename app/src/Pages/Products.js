import React, { useEffect, useState, useReducer } from "react";
import PropTypes from "prop-types";

/* Controller */
import { search } from "../../server/controllers/products";

/* Bootstrap */
import { Row, Col } from "react-bootstrap";

/* Currency Helper */
import { formatAsCurrency } from "../../server/utils/helpers";

/* Routing */
import { Link } from "react-router-dom";

/* Personal components */
import Breadcrumb from "../Components/Breadcrumb";
import Message from "../Components/Message";

/* Loading */
import Loading from "../Components/Loading";

/* Reducer products  */
import { productsReducer, initialState } from "../../reducers/products";
import { FETCH_SUCCESS, FETCH_ERROR } from "../../reducers/actions/products";
const Products = ({ searchItem }) => {
  const [{ products, error, loading }, dispatch] = useReducer(
    productsReducer,
    initialState
  );
  const [categories, setCategories] = useState([]);
  const fetchData = async () => {
    try {
      const { items, categories } = await search(searchItem);
      setCategories(categories);
      dispatch({ type: FETCH_SUCCESS, payload: items });
    } catch (e) {
      dispatch({ type: FETCH_ERROR, payload: items });
    }
  };
  useEffect(() => {
    fetchData();
  }, [searchItem]);

  const RenderProducts = () => {
    return (
      <React.Fragment>
        <Breadcrumb categories={categories} />
        <section>
          {products.map(
            ({ id, title, price, free_shipping, address, picture }) => (
              <Row
                key={id}
                className="justify-content-center bg-white pb-16 pt-16 pl-32 first-row"
              >
                <Col md={9}>
                  <Row>
                    <Col md={3} className="thumbnail">
                      <Link to={`/product/${id}`}>
                        <picture>
                          {/* Formatos */}
                          <img
                            className="pointer rounded thumbnail"
                            src={picture}
                            alt={title}
                          />
                        </picture>
                      </Link>
                    </Col>
                    <Col md={9}>
                      <h4 className="f-24">
                        $ {formatAsCurrency(price.amount)}{" "}
                        {free_shipping ? (
                          <img
                            src="https://images-prod-meli.s3-sa-east-1.amazonaws.com/ic_shipping.png"
                            alt="free shipping"
                          />
                        ) : null}
                      </h4>

                      <Link
                        to={`/product/${id}`}
                        style={{
                          color: "black",
                          textDecoration: "inherit",
                        }}
                      >
                        <p className="pointer f-18 mt-32">{title}</p>
                      </Link>
                    </Col>
                  </Row>
                </Col>
                <Col md={3} className="pt-16 pl-16">
                  <p className="f-12"> {address}</p>
                </Col>
              </Row>
            )
          )}
        </section>
      </React.Fragment>
    );
  };

  return (
    <React.Fragment>
      {/* Se muestra si hay productos y loading = false */}
      {loading ? <Loading /> : null}
      {!!products.length ? (
        <RenderProducts />
      ) : (
        <Message message={"Buscar productos ..."} />
      )}
      {error ? <h3>Ocurri√≥ un error, intente de nuevo</h3> : null}
    </React.Fragment>
  );
};

Products.propTypes = {
  searchItem: PropTypes.string,
};
export default Products;
