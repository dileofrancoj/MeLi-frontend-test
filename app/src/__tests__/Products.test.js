import React from "react";
import { shallow } from "enzyme";
import { StaticRouter } from "react-router-dom";
import { search } from "../../server/controllers/products";
import Products from "./../Pages/Products";

describe("Products", () => {
  test("should render basic Products component ", () => {
    const component = shallow(
      <StaticRouter>
        <Products />
      </StaticRouter>
    );
    expect(component).toMatchSnapshot();
  });

  test("should render Products with a prop ", () => {
    const component = shallow(
      <StaticRouter>
        <Products searchItem="Macbook" />
      </StaticRouter>
    );
    expect(component).toMatchSnapshot();
  });

  test("should get an Object", async () => {
    const component = shallow(
      <StaticRouter>
        <Products searchItem="Macbook" />
      </StaticRouter>
    );
    const product = "macbook";
    const data = await search(product);
    expect(typeof data).toBe("object");
  });
});
