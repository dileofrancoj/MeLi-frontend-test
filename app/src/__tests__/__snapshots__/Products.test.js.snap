// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products should render Products with a prop  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <Products
    searchItem="Macbook"
  />
</Router>
`;

exports[`Products should render basic Products component  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <Products />
</Router>
`;
